target/ppc/fpu_helper.c

```diff
@@ -3368,84 +3368,199 @@ static inline target_ulong compute_fp_flags(float_status *excp_ptr)
+            excp_ptr = NULL;                                            \
+        }                                                               \
         b = cpu_vsr_ptr(env, b_r);                                      \
-        excp_ptr = &env->fp_status;                                     \
         fpscr = 0;                                                      \
-        acc = (bool)(packed_flags & GER_ACC); /* Accumulate */          \
-        neg_acc = (bool)(packed_flags & GER_NA); /* Neg Accumulate */   \
-        neg_mul = (bool)(packed_flags & GER_NM); /* Neg Multiply */     \
         for (i = 0, xmsk_bit = 1 << 3; i < 4; i++, xmsk_bit >>= 1) {    \
             a = cpu_vsr_ptr(env, a_r + i/ymax);                         \
             at = cpu_vsr_ptr(env, at_r + i);                            \
             for (j = 0, ymsk_bit = 1 << (ymax-1); j < ymax;             \
                  j++, ymsk_bit >>= 1) {                                 \
                 if ((xmsk_bit & xmsk) && (ymsk_bit & ymsk)) {           \
-                    helper_reset_fpstatus(env);                         \
+                    if (FLOAT) {                                        \
+                        helper_reset_fpstatus(env);                     \
+                    }                                                   \
                     psum = (TARGET_T)0;                                 \
                     for (k = 0, pmsk_bit = 1 << (RANK-1); k < RANK;     \
                          k++, pmsk_bit >>= 1) {                         \
                         if (pmsk_bit & pmsk) {                          \
-                            va = GET_VSR(Vsr##OR_EL, a, RANK*(i%ymax) + \
-                                         k, ORIG_T, TARGET_T);          \
-                            vb = GET_VSR(Vsr##OR_EL, b, RANK*j + k,     \
-                                         ORIG_T, TARGET_T);             \
+                            if (FLOAT) {                                \
+                                va = GET_VSR(Vsr##OR_EL, a, RANK*       \
+                                        (i%ymax) + k, ORIG_T, TARGET_T);\
+                            } else {                                    \
+                                va = GET_VSR(VsrS##OR_EL, a, RANK*i + k,\
+                                             ORIG_T, TARGET_T);         \
+                            }                                           \
+                            if (!FLOAT && (RANK == 2 || RANK == 8)) {   \
+                                vb = GET_VSR(VsrS##OR_EL, b, RANK*j + k,\
+                                             ORIG_T, TARGET_T);         \
+                            } else {                                    \
+                                vb = GET_VSR(Vsr##OR_EL, b, RANK*j + k, \
+                                             ORIG_T, TARGET_T);         \
+                            }                                           \
```

Comment:
I tried to avoid overusing if(FLOAT) statements, but some (specially related to floating point exceptions) I wasn't able to solve without it. 
In this case the problem was dealing with the fact that almost all int Vsr calls are signed, but xvi8ger4[pp] has a unsigned int Vsr, so the 'S' couldn't be part of the GET_Vsr macro for integers.